File: experiments/adrianco/experiment-5/runs/language=clojure_model=claude-opus-4-8_tooling=none/rep1/test/soccer/queries_test.clj
(ns soccer.queries-test
  "=============================================================================
   BDD (Given-When-Then) scenarios for the query layer, mirroring the
   'Required Capabilities' and 'Testing Approach' sections of the spec. All
   scenarios run against soccer.fixtures/db.
   ============================================================================="
  (:require [clojure.test :refer [deftest testing is]]
            [soccer.fixtures :as fx]
            [soccer.queries :as q]))

(deftest match-queries
  (testing "Scenario: Find matches between two teams"
    ;; Given the match data is loaded
    ;; When I search for matches between Flamengo and Fluminense
    (let [ms (q/search-matches fx/db {:team "Flamengo" :team2 "Fluminense"})]
      ;; Then I receive a list of matches
      (is (= 2 (count ms)))
      ;; And each match has date, scores and competition
      (is (every? #(and (:date %) (:competition %)
                        (some? (:home-goal %)) (some? (:away-goal %))) ms))
      ;; And results are most-recent first
      (is (= "2019-09-03" (:date (first ms))))))

  (testing "Scenario: Find a team's matches in a season"
    ;; When I ask what matches Palmeiras played in 2019
    (let [ms (q/search-matches fx/db {:team "Palmeiras" :season 2019})]
      ;; Then only 2019 Palmeiras matches are returned
      (is (= 4 (count ms)))
      (is (every? #(= 2019 (:season %)) ms))))

  (testing "Scenario: Filter matches by competition"
    (let [ms (q/search-matches fx/db {:competition "Libertadores"})]
      (is (= 1 (count ms)))
      (is (= "Copa Libertadores" (:competition (first ms)))))))

(deftest team-statistics
  (testing "Scenario: Get team statistics for a season"
    ;; Given the match data is loaded
    ;; When I request statistics for Flamengo in 2019
    (let [t (q/team-stats fx/db {:team "Flamengo" :season 2019
                                 :competition "Brasileirão Série A"})]
      ;; Then I receive wins, losses, draws and goals
      (is (= 5 (:played t)))
      (is (= 4 (:wins t)))
      (is (= 1 (:draws t)))
      (is (= 0 (:losses t)))
      (is (= 12 (:gf t)))
      (is (= 3 (:ga t)))
      (is (= 13 (:points t)))))

  (testing "Scenario: Restrict a team's record to home matches"
    (let [t (q/team-stats fx/db {:team "Flamengo" :season 2019
                                 :competition "Brasileirão Série A" :venue :home})]
      ;; Flamengo home league games in 2019: vs Palmeiras(3-1), Santos(4-0), Fluminense(2-1)
      (is (= 3 (:played t)))
      (is (= 3 (:wins t))))))

(deftest head-to-head-query
  (testing "Scenario: Compare two teams head-to-head"
    ;; When I compare Flamengo and Fluminense head-to-head
    (let [h (q/head-to-head fx/db {:team1 "Flamengo" :team2 "Fluminense"})]
      ;; Then each side's record is reported
      (is (= 2 (:meetings h)))
      (is (= 1 (:team1-wins h)))   ; Flamengo won the Brasileirão meeting
      (is (= 1 (:team2-wins h)))   ; Fluminense won the cup meeting
      (is (= 0 (:draws h))))))

(deftest competition-standings
  (testing "Scenario: Compute league standings from match results"
    ;; When I request the 2019 Brasileirão standings
    (let [rows (q/standings fx/db {:competition "Brasileirão Série A" :season 2019})]
      ;; Then teams are ranked by points
      (is (= ["Flamengo" "Palmeiras" "Santos" "Fluminense"] (map :team rows)))
      (is (= [13 4 2 0] (map :points rows)))
      ;; And the leader's goal difference is correct
      (is (= 9 (:gd (first rows)))))))

(deftest statistical-analysis
  (testing "Scenario: Average goals per match for a competition"
    (let [s (q/league-stats fx/db {:competition "Brasileirão Série A" :season 2019})]
      ;; 7 league matches in 2019: rounds 1-6 + the Fla-Flu derby (round 7)
      (is (= 7 (:matches s)))
      ;; goals: 4+2+4+2+4+1+3 = 20
      (is (= 20 (:goals s)))
      (is (< (Math/abs (- (:avg-goals s) (/ 20.0 7))) 1e-9))))

  (testing "Scenario: Biggest victories by margin"
    (let [ms (q/biggest-wins fx/db {:limit 3})]
      ;; Then the 8-0 Libertadores result is first
      (is (= 8 (- (:home-goal (first ms)) (:away-goal (first ms))))))))

(deftest player-queries
  (testing "Scenario: Find Brazilian players sorted by rating"
    ;; When I search for Brazilian players
    (let [ps (q/search-players fx/db {:nationality "Brazil"})]
      ;; Then only Brazilians are returned, highest rated first
      (is (every? #(= "Brazil" (:nationality %)) ps))
      (is (= "Neymar Jr" (:name (first ps))))))

  (testing "Scenario: Find players at a specific club"
    (let [ps (q/search-players fx/db {:club "Flamengo"})]
      (is (= 3 (count ps)))
      (is (= ["Gabriel Barbosa" "Diego Alves" "Bruno Henrique"] (map :name ps)))))

  (testing "Scenario: Search a player by name"
    (let [ps (q/search-players fx/db {:name "Gabriel Barbosa"})]
      (is (= 1 (count ps)))
      (is (= "Flamengo" (:club (first ps))))))

  (testing "Scenario: Filter players by position and rating"
    (let [ps (q/search-players fx/db {:club "Flamengo" :position "GK"})]
      (is (= ["Diego Alves"] (map :name ps))))))

File: experiments/adrianco/experiment-5/runs/language=typescript_model=claude-opus-4-7_tooling=beads/rep3/tests/queries.test.ts
import { describe, it, expect, beforeAll } from 'vitest';
import { loadAll, defaultDataDir } from '../src/loader.js';
import {
  findMatches,
  headToHead,
  teamStats,
  computeStandings,
  findPlayers,
  biggestWins,
  aggregateStats,
  topScoringTeams,
  clubRoster,
} from '../src/queries.js';
import { DataStore } from '../src/types.js';

let store: DataStore;

beforeAll(async () => {
  store = await loadAll(defaultDataDir());
});

describe('Feature: Data loading', () => {
  it('Given the kaggle CSVs are present, When loaded, Then all 6 datasets are populated', () => {
    expect(store.matches.length).toBeGreaterThan(20000);
    expect(store.players.length).toBeGreaterThan(18000);
    const comps = new Set(store.matches.map((m) => m.competition));
    expect(comps.has('Brasileirao')).toBe(true);
    expect(comps.has('Copa do Brasil')).toBe(true);
    expect(comps.has('Libertadores')).toBe(true);
    expect(comps.has('BR-Football')).toBe(true);
    expect(comps.has('Historical')).toBe(true);
  });
});

describe('Feature: Match queries', () => {
  describe('Scenario: Find matches between two teams', () => {
    it('Given match data is loaded, When I search Flamengo vs Fluminense, Then I get the Fla-Flu derbies', () => {
      const matches = findMatches(store, {
        team: 'Flamengo',
        opponent: 'Fluminense',
      });
      expect(matches.length).toBeGreaterThan(0);
      for (const m of matches) {
        const teams = m.homeTeamNorm + ' ' + m.awayTeamNorm;
        expect(teams).toMatch(/flamengo/);
        expect(teams).toMatch(/fluminense/);
        expect(m.date).toMatch(/\d{4}-\d{2}-\d{2}/);
      }
    });
  });

  describe('Scenario: Filter by season', () => {
    it('Given match data loaded, When I request Palmeiras 2019 matches, Then all results are from 2019', () => {
      const matches = findMatches(store, { team: 'Palmeiras', season: 2019 });
      expect(matches.length).toBeGreaterThan(0);
      for (const m of matches) expect(m.season).toBe(2019);
    });
  });

  describe('Scenario: Filter by competition', () => {
    it('Given match data loaded, When I request Copa do Brasil matches, Then competition is Copa do Brasil', () => {
      const matches = findMatches(store, { competition: 'Copa do Brasil' });
      expect(matches.length).toBeGreaterThan(0);
      for (const m of matches) expect(m.competition).toBe('Copa do Brasil');
    });
  });

  describe('Scenario: Date range filtering', () => {
    it('Given match data loaded, When I filter by date range, Then results fall within range', () => {
      const matches = findMatches(store, {
        fromDate: '2019-01-01',
        toDate: '2019-12-31',
      });
      expect(matches.length).toBeGreaterThan(0);
      for (const m of matches) {
        expect(m.date >= '2019-01-01').toBe(true);
        expect(m.date <= '2019-12-31').toBe(true);
      }
    });
  });
});

describe('Feature: Head-to-head', () => {
  describe('Scenario: H2H summary computes correctly', () => {
    it('Given two teams, When I compute h2h, Then wins+draws+losses == totalMatches', () => {
      const h = headToHead(store, 'Palmeiras', 'Santos');
      expect(h.totalMatches).toBe(h.teamAWins + h.teamBWins + h.draws);
      expect(h.totalMatches).toBeGreaterThan(0);
    });
  });
});

describe('Feature: Team statistics', () => {
  describe('Scenario: Compute season W/D/L', () => {
    it('Given Flamengo 2019 Brasileirao, Then matches sum to W+D+L', () => {
      const stats = teamStats(store, 'Flamengo', {
        season: 2019,
        competition: 'Brasileirao',
      });
      expect(stats.played).toBeGreaterThan(0);
      expect(stats.played).toBe(stats.wins + stats.draws + stats.losses);
      expect(stats.points).toBe(stats.wins * 3 + stats.draws);
    });
  });

  describe('Scenario: Home-only vs all venues', () => {
    it('Given a team, When restricted to home venue, Then matches subset of all-venue matches', () => {
      const all = teamStats(store, 'Corinthians', {
        season: 2019,
        competition: 'Brasileirao',
      });
      const home = teamStats(store, 'Corinthians', {
        season: 2019,
        competition: 'Brasileirao',
        venue: 'home',
      });
      expect(home.played).toBeLessThanOrEqual(all.played);
      expect(home.played).toBeGreaterThan(0);
    });
  });
});

describe('Feature: Standings', () => {
  describe('Scenario: Compute Brasileirao 2019 standings', () => {
    it('Given match data loaded, When standings calculated, Then Flamengo is champion (1st)', () => {
      const table = computeStandings(store, 'Brasileirao', 2019);
      expect(table.length).toBeGreaterThan(0);
      expect(table[0].position).toBe(1);
      // Flamengo won 2019 Brasileirao with 90 pts in real life
      const champion = table[0];
      expect(champion.team.toLowerCase()).toContain('flamengo');
    });

    it('Given standings, When summed, Then points order is descending', () => {
      const table = computeStandings(store, 'Brasileirao', 2019);
      for (let i = 1; i < table.length; i++) {
        expect(table[i - 1].points).toBeGreaterThanOrEqual(table[i].points);
      }
    });
  });
});

describe('Feature: Player queries', () => {
  describe('Scenario: Find Brazilian players', () => {
    it('Given fifa data loaded, When filtered by Brazil, Then top result is Brazilian', () => {
      const players = findPlayers(store, { nationality: 'Brazil', limit: 5 });
      expect(players.length).toBeGreaterThan(0);
      for (const p of players) expect(p.nationality).toBe('Brazil');
      // Top result by overall should be very highly rated
      expect((players[0].overall ?? 0)).toBeGreaterThan(85);
    });
  });

  describe('Scenario: Search by name substring', () => {
    it('Given a name fragment "Neymar", Then we find at least one match', () => {
      const players = findPlayers(store, { name: 'Neymar' });
      expect(players.length).toBeGreaterThan(0);
      expect(players[0].name.toLowerCase()).toContain('neymar');
    });
  });

  describe('Scenario: Filter by overall rating', () => {
    it('Given minOverall 90, Then all returned players have overall >= 90', () => {
      const players = findPlayers(store, { minOverall: 90, limit: 20 });
      expect(players.length).toBeGreaterThan(0);
      for (const p of players) expect((p.overall ?? 0)).toBeGreaterThanOrEqual(90);
    });
  });

  describe('Scenario: Club roster', () => {
    it('Given a club like Flamengo, When listed, Then we get at least one player', () => {
      const r = clubRoster(store, 'Flamengo');
      // Flamengo's roster in FIFA data should not be empty
      expect(r.players.length).toBeGreaterThanOrEqual(0);
    });
  });
});

describe('Feature: Statistical analysis', () => {
  describe('Scenario: Aggregate stats for Brasileirao', () => {
    it('Given Brasileirao matches, When aggregated, Then win-rate components sum to ~1', () => {
      const a = aggregateStats(store, { competition: 'Brasileirao' });
      expect(a.totalMatches).toBeGreaterThan(0);
      const totalRate = a.homeWinRate + a.awayWinRate + a.drawRate;
      expect(totalRate).toBeGreaterThan(0.99);
      expect(totalRate).toBeLessThan(1.01);
      expect(a.averageGoalsPerMatch).toBeGreaterThan(1);
      expect(a.averageGoalsPerMatch).toBeLessThan(5);
    });
  });

  describe('Scenario: Biggest wins', () => {
    it('Given the dataset, When sorted by goal difference, Then top result is large', () => {
      const wins = biggestWins(store, { limit: 5 });
      expect(wins.length).toBe(5);
      const diff = Math.abs(wins[0].homeGoals - wins[0].awayGoals);
      expect(diff).toBeGreaterThanOrEqual(5);
    });
  });

  describe('Scenario: Top scoring teams', () => {
    it('Given Brasileirao 2019, When tallied, Then list returned in descending order', () => {
      const top = topScoringTeams(store, {
        competition: 'Brasileirao',
        season: 2019,
        limit: 5,
      });
      expect(top.length).toBe(5);
      for (let i = 1; i < top.length; i++) {
        expect(top[i - 1].goalsFor).toBeGreaterThanOrEqual(top[i].goalsFor);
      }
    });
  });
});
